### DISCUSSION TITLE
Right SFP EEPROM config for DFP-34X-2C2 to work automatically work at 2500base-X?

### DISCUSSION THREAD

**Discussion by Headcrabed** [Anime4000/RTL960x#250] [Ideas]
Edit: This procedure also works for ODI DFP-34X-2I3, more details [here](https://forum.banana-pi.org/t/bpi-r4-and-sfp/16945/182)

Seems ODI has wrong configuration in DFP-34X-2C2's eeprom. And this blocks linux to use 2500base-X by default.

First, they set BR, Nominal value to 1300MBd, while it should be 3125MBd(Since 3125 cannot be set, it should be round off to 3100MBd).

Second, 1000base-LX is enabled by default, so that linux would prefer that speed (https://github.com/torvalds/linux/blob/cf1182944c7cc9f1c21a8a44e0d29abe12527412/drivers/net/phy/sfp-bus.c#L327)

After disabling 1000base-LX and set BR, Nominal value to 3100MBd in EEPROM, my device (Banana pi BPI-R3 + Openwrt 23.05.2) would use 2500base-X by default now.

For example:
![图片](https://github.com/Anime4000/RTL960x/assets/56786900/63a461cd-a28f-4305-8116-0c712bbf79c7)
Set addr 0x06 to 0x00(Disable 1000base-LX), addr 0x0c to 0x1f(3100MBd), addr 0x3f to 0x80(Checksum)

When using openwrt, you need to install ethtool-full and i2c-tools before operating. To dump eeprom, you can use ```ethtool -m *your sfp port here* raw on > dump.bin```, then download the bin using winscp or something and open it using SFPTotal Wizard(It is used to calculate checksum, you can use other software instead)

**Comment by Anime4000** [comment]
to change this, do you use SFPTotal Whistle or Mini ?

**Reply by Headcrabed** [reply]
I didn’t use any of them. I use SFPTotal Wizard just to make it easier to calculate checksums.

For editing eeprom, I install i2c-tools on my router and directly use i2cset to edit EEPROM.

**Reply by Anime4000** [reply]
can you show example of i2cset usage?

**Reply by Headcrabed** [reply]
On my device: i2cset -y 1(number of i2c bus,3 total on my device) 0x50(eeprom’s i2c address) 0x06(Target address in eeprom) 0x00 (target value)

**Reply by Anime4000** [reply]
I see, if to change other block by simple execute multiple time (different target address and new target value?)

**Reply by Headcrabed** [reply]
Yes, don't forget to change checksum value(Green blocks inside the picture above)

**Reply by Anime4000** [reply]
if using raspberry pi 3 i2c to access SFP i2c, it should be possible? can power sfp just from pi 3 or need external power?

**Reply by Headcrabed** [reply]
I didn't try this, but I think that's not enough, It's better to try it by yourself 😑

**Reply by Headcrabed** [reply]
Besides, remember that you can only set 1 speed mode, since sfp spec doesn't support speed auto negotiation if using real eeprom rather than fake ones.

**Reply by Headcrabed** [reply]
@Anime4000 btw your description in readme.md is quite misleading. The actual situation is that any Linux host without hack would only work at 1000baselx mode, and modify eeprom is the only method to switch to 2500basex unless adding a quirk to Linux driver. 
However mainline refused to accept the quirk for this stick as it has the possibility to lock user out(wrong LAN_SDS_MODE)

**Reply by Anime4000** [reply]
@Headcrabed so I use this?
```
Linux default to 1000base-LX per EEPROM, but mainline Linux disregards LAN_SDS_MODE quirks.
```

**Reply by Headcrabed** [reply]
What about this: “Linux host need this for 2.5gbps mode if not hacking kernel”

**Reply by Anime4000** [reply]
sounds better and simple!

**Reply by Anime4000** [reply]
@Headcrabed I have question, that mode support which Linux Kernel version?

**Reply by Headcrabed** [reply]
@Anime4000 Emmm.. What do you mean? If you mean linux support of 2500base-x on sfp module, seems it has been supported at least since 4.16-rc1: https://github.com/torvalds/linux/commit/f10fcbcf91006a3223cffbc413b5abe17cf46937

**Reply by Anime4000** [reply]
I mean: Disable 1000base-LX + 0x1f (3100MBd), refering to git blame, the change was 6 month ago, I wonder this method support older Linux Kernel

**Reply by Headcrabed** [reply]
That's 6 years ago.. And I didn't find code for 3100MBd support on earlier versions of kernel.

**Reply by Anime4000** [reply]
based on this change:
https://github.com/torvalds/linux/commit/466b97b1871a49c272757f7527db2f85ea6338b4
it committed on Nov 17, 2023

code change before `if (bitmap_empty(modes, __ETHTOOL_LINK_MODE_MASK_NBITS) && br_nom)` should give same results?

**Reply by Headcrabed** [reply]
I think it should behave the same before and after the commit you mentioned

**Reply by Anime4000** [reply]
@Headcrabed accroding to SFF-8472, at Addr 0x0a (Index 10) can identify fiber speed, I saw many DAC cable and 10G RJ45 SFP has this value filled `0x67` (0110 0111),

if this stick set to `0x05` (0000 0101) like this, OpenWRT will auto nego as well?
| Addr | Value |
|------|-------|
| 0x06 | 0x00  |
| 0x0a | 0x05  |
| 0x0c | 0x1f  |

![Untitled-1](https://github.com/user-attachments/assets/e1919eda-f32f-4619-bbbf-5ab77d7ca7fc)

**Reply by Headcrabed** [reply]
> @Headcrabed accroding to SFF-8472, at Addr 0x0a (Index 10) can identify fiber speed, I saw many DAC cable and 10G RJ45 SFP has this value filled `0x67` (0110 0111),
> 
> if this stick set to `0x05` (0000 0101) like this, OpenWRT will auto nego as well?
> Addr 	Value
> 0x06 	0x00
> 0x0a 	0x05
> 0x0c 	0x1f
> 
> ![Untitled-1](https://private-user-images.githubusercontent.com/1908715/377014105-e1919eda-f32f-4619-bbbf-5ab77d7ca7fc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjkwOTQwNDUsIm5iZiI6MTcyOTA5Mzc0NSwicGF0aCI6Ii8xOTA4NzE1LzM3NzAxNDEwNS1lMTkxOWVkYS1mMzJmLTQ2MTktYmJiZi01YWI3N2Q3Y2E3ZmMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMTZUMTU0OTA1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YjE0MjI0YWNiZWQzYjgyZjMxMTJlOGM0OWQzNTE1ZWRkOGRkMzgyMzMwMGExNzIyMTQwN2ZmMDNhNTQwZTJjMSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.JqFSZjGqJA_vfTY3bJdP7M48XiQdbNPMGw4bfExTK3s)

I don't think so. That's a different speed mode, not 2.5gbps,

**Reply by Anime4000** [reply]
I see, I try play around that address LSB value, it did represent this:
![image](https://github.com/user-attachments/assets/6ae64912-e483-4b75-855a-5f450a544a32)

there is another way to tell 2500Base-X ?

**Comment by duskyzaoKING** [comment]
Can you share the .bin of the entire firwmare with this modification ? will use a programer to flash direct in SOP

**Reply by Headcrabed** [reply]
Ok, I will do it later. But I still prefer you to do the modification by yourself, or you will lose your date code and serial number.

**Reply by duskyzaoKING** [reply]
Didnt thinked about it. I think its best i do. thanks

**Reply by Headcrabed** [reply]
[dump_eeprom.zip](https://github.com/Anime4000/RTL960x/files/15282616/dump_eeprom.zip)
This is the dump of addr 0x50 of my eeprom

**Comment by lehoangnb** [comment]
I found this command can change EEPROM value, but eeprom content does not same as I read from chip using programer, anyone can check it work or not?
`diag i2c set port 1 dev 0x50 reg 0x06 data 0x0`
(For checksum, use `diag i2c get port 1 dev 0x50 reg ...` to dump all eeprom content from 0x00 to 0x3E and use SFPTotal to calculate)

**Reply by Anime4000** [reply]
I tried this, diag i2c wont write to it, I wonder if there easy way to write internally

**Reply by lehoangnb** [reply]
Yes, I think its is dummy eeprom, the value remain after reboot but doesn't same as real eeeprom
I have dump all value from i2c diag
![image](https://github.com/Anime4000/RTL960x/assets/5148339/7857f5b4-05d3-44d5-8b9e-10cd38fe328a)

**Comment by Anime4000** [comment]
I tried 
| Addr | Value |
|------|-------|
| 0x06 | 0x00  |
| 0x0c | 0x1F  |

Mikrotik detect as this instance
![image](https://github.com/Anime4000/RTL960x/assets/1908715/8cc4a44c-00c3-43d9-8f2f-b576a62eec5c)

The way Mikrotik show supported are different, I wonder what value to make `2.5G baseX` in supported list

**Reply by Headcrabed** [reply]
Can you share a full dump of eeprom? Raw version.

**Reply by lehoangnb** [reply]
Follow source of ethtool, no value for 2.5G
https://github.com/serhepopovych/ethtool/blob/master/sfpid.c#L97

**Reply by Anime4000** [reply]
@Headcrabed here
[NIJK24000001.zip](https://github.com/user-attachments/files/15528513/NIJK24000001.zip)

@lehoangnb 
maybe buy 2.5G/1G RJ45 SFP and copy the EEPROM see what position is use

**Reply by Headcrabed** [reply]
> @Headcrabed here [NIJK24000001.zip](https://github.com/user-attachments/files/15528513/NIJK24000001.zip)
> 
> @lehoangnb maybe buy 2.5G/1G RJ45 SFP and copy the EEPROM see what position is use

Seems everything is right. So that's Mikrotik's fault if you are sure yore device supports 2.5G mode.

**Reply by Anime4000** [reply]
if buy 2.5/1G RJ45 SFP and copy the EEPROM would work? but it become 2.5G baseT

@Headcrabed 
    2.5G just work by disable auto nego and set to 2.5G baseX, I tried 2.5G baseT also works. I just wondering if SFP Support list show 2.5G baseX and can use auto nego

**Reply by Headcrabed** [reply]
That's Mikrotik's fault to not supporting auto negotiation.

**Reply by Anime4000** [reply]
if I buy RJ45 2.5G/1G and copy it's EEPROM, will it work?

**Reply by Headcrabed** [reply]
Nope, it won't help.

**Reply by Anime4000** [reply]
@Headcrabed I managed to add `2.5G baseX` in **SFP Support** list
![image](https://github.com/user-attachments/assets/f9446ee1-5b74-4dcf-b24d-d6afeb4207af)

pair with Ethernet Compliance Codes `0x06`=`0x00` (Undefined) and BR Nominal `0x0c`=`0x1f` (31) it will display `10G baseSR LR`, I think this is bug perhaps?

**Reply by Headcrabed** [reply]
If you are sure you didn't modify anything except vendor info and have already updated checksum info(There are 2 different checksums for different parts of the eeprom), it should be your ROS' bug.

**Reply by Anime4000** [reply]
Checksum is good, 
I talked to others indeed ROS' bug, what I test:
0x06 = 0x02 & 0x0c = 0x1f prints `10G baseSR LR` together with `2.5G baseX`
0x06 = 0x02 & 0x0c = 0x19 doesn't print `10G baseSR LR` and `2.5G baseX`
0x06 = 0x00 & 0x0c = 0x0d doesn't print `10G baseSR LR` and `2.5G baseX`

so, the Addr 0x0c >= 0x1f (>=3100) make ROS think is a `10G baseSR LR`

**Reply by vuducdong** [reply]
> Checksum is good, I talked to others indeed ROS' bug, what I test: 0x06 = 0x02 & 0x0c = 0x1f prints `10G baseSR LR` 0x06 = 0x02 & 0x0c = 0x19 doesn't print `10G baseSR LR` 0x06 = 0x00 & 0x0c = 0x0d doesn't print `10G baseSR LR`
> 
> so, the Addr 0x0c >= 0x1f (>=3100) make ROS think is a `10G baseSR LR`

Please send report it to Mikrotik technical support.

**Reply by Anime4000** [reply]
> > Checksum is good, I talked to others indeed ROS' bug, what I test: 0x06 = 0x02 & 0x0c = 0x1f prints `10G baseSR LR` 0x06 = 0x02 & 0x0c = 0x19 doesn't print `10G baseSR LR` 0x06 = 0x00 & 0x0c = 0x0d doesn't print `10G baseSR LR`
> > so, the Addr 0x0c >= 0x1f (>=3100) make ROS think is a `10G baseSR LR`
> 
> Please send report it to Mikrotik technical support.

Forum or on this page?
![image](https://github.com/user-attachments/assets/40a80039-0120-4114-950f-8fe1c449d426)

**Reply by vuducdong** [reply]
> > > Checksum is good, I talked to others indeed ROS' bug, what I test: 0x06 = 0x02 & 0x0c = 0x1f prints `10G baseSR LR` 0x06 = 0x02 & 0x0c = 0x19 doesn't print `10G baseSR LR` 0x06 = 0x00 & 0x0c = 0x0d doesn't print `10G baseSR LR`
> 
> > > so, the Addr 0x0c >= 0x1f (>=3100) make ROS think is a `10G baseSR LR`
> 
> > 
> 
> > Please send report it to Mikrotik technical support.
> 
> 
> 
> Forum or on this page?
> 
> ![image](https://github.com/user-attachments/assets/40a80039-0120-4114-950f-8fe1c449d426)
> 
> 

Yes, this page

**Reply by Anime4000** [reply]
Alright, I need to craft the support and detailed documentation, I use MA5671a EEPROM instead since can be edit via SSH

**Comment by ifplusor** [comment]
I test this on [Banana Pi BPI-R4](https://wiki.banana-pi.org/Banana_Pi_BPI-R4), it's work.
![image](https://github.com/Anime4000/RTL960x/assets/9999114/9731b8b5-25a4-4e81-ab72-c567765b6f74)

**Comment by Hannabiann** [comment]
How to flashing modified firmware? Only programmer?

**Reply by Headcrabed** [reply]
Please check the discussion on 1st reply
